### PR TITLE
[NVPTX] Add tex.grad.cube{array} intrinsics

### PR DESCRIPTION
**Pull request by AlexMaclean**
Extend IR support for PTX `tex` instruction described in [PTX ISA. 9.7.9.3. Texture Instructions: tex](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#texture-instructions-tex). Add support for unified-move versions of `tex.grad.cube{array}` variants added in PTX ISA 4.3.

### CODE REVIEW

**Comment by AlexMaclean** [comment]
@Artem-B @ThomasRaoux please review when you have a chance.

**Comment by Artem-B** [comment]
The code looks OK, but it would be great to see the tests verifying that the new instructions are generated correctly.

**Comment by AlexMaclean** [comment]
@Artem-B, the changes to `llvm/test/CodeGen/NVPTX/surf-tex.py` add coverage for these intrinsics, is there additional testing you'd like?

**Comment by Artem-B** [comment]
> @Artem-B, the changes to `llvm/test/CodeGen/NVPTX/surf-tex.py` add coverage for these intrinsics, is there additional testing you'd like?

Ah. It was not obvious. Can you please verify that the test does contain the new instructions and that LLVM tests pass with `LLVM_PTXAS_EXECUTABLE=/path/to/ptxas` set in the environment.

File: llvm/test/CodeGen/NVPTX/surf-tex.py [hunk 1]
```diff
@@ -1,12 +1,12 @@
 # RUN: %python %s --target=cuda --tests=suld,sust,tex,tld4 --gen-list=%t.list > %t-cuda.ll
-# RUN: llc -mcpu=sm_20 %t-cuda.ll -verify-machineinstrs -o - | FileCheck %t-cuda.ll
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/surf-tex.py:2]: We should still have the CPU set explicitly. `sm_60` would probably be the reasonable oldest non-deprecated arch.

File: llvm/test/CodeGen/NVPTX/surf-tex.py [hunk 2]
```diff
@@ -353,8 +364,9 @@ def gen_sust_tests(target, global_surf):
     ret void
   }
   ; CHECK-LABEL: .entry ${test_name}_global
-  ; CHECK: ${instruction} [${global_surf}, ${reg_access}], ${reg_value}
-  ;
+  ; CHECK-CUDA: mov.u64 [[REG${reg_id}:%.*]], ${global_surf}
+  ; CHECK-CUDA: ${instruction} [[[REG${reg_id}]], ${reg_access}], ${reg_value}
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/surf-tex.py:368]: I wonder why we end up with an extra `mov` for CUDA, but not openCL. 
The issue is cosmetic, and is not a blocker for this patch.

**Review by Artem-B — APPROVED** [review]
LGTM w/ a minor test nit.

File: llvm/test/CodeGen/NVPTX/surf-tex.py (hunk 1 above)
Comment (AlexMaclean) [reply llvm/test/CodeGen/NVPTX/surf-tex.py:2]: Sounds good, fixed.

**Review by Artem-B — APPROVED** [review]

**Comment by AlexMaclean** [comment]
@Artem-B, could you please land these changes on my behalf?

**Comment by AlexMaclean** [comment]
@Artem-B, ping

**Pull request merged**